### PR TITLE
feat(@angular-devkit/build-angular): differential loading check

### PR DESCRIPTION
**Pull request by manfredsteyer**

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  const target = getTargetVersion(tsConfigFileName);
+
+  return es5SupportNeeded && target !== ScriptTarget.ES5;
+}
+
+export function getTargetVersion(tsConfigFileName: string): ScriptTarget {
+  return _getTargetVersion(tsConfigFileName, new Set<string>());
+}
+
+function _getTargetVersion(tsConfigFileName: string, visited: Set<string>): ScriptTarget {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: We shouldn't be re-creating the ts parsing parsing logic of typescript. You can simply use `readTsconfig` method.

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
+/**
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Can you please add return types to methods?

Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Maybe this file is better names as `utils.ts`, what do you think?

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    return _getTargetVersion(nextTsConfigFileName, visited);
+  } else {
+    // Fallback to ES5
+    return ScriptTarget.ES5;
+  }
+
+}
+
+export function isEs5BrowserSupportNeeded(
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Some docs would help here. Example what is a `query`?

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+function convertSupportTableToBrowserVersionTuples(supportTable: object): Map<string, string> {
+  // The supportTable bases upon data from the caniuse api.
+  // It uses the y property to indicate the version beginning
+  // with the feature in question is supported.
+  const browsers = Object.getOwnPropertyNames(supportTable) as string[];
+  const supportMap = new Map<string, string>();
+
+  browsers
+     // @ts-ignore
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Why is this needed?

File: packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
+/**
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts]: This seems to be a `spec` file which can be combined with the others.

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 4 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: We don't have typings for the data structures provided by caniuse. Hence, here it's just an object (supportTable: object). However, `supportTable[b].y` leads to an implicit any issue.

File: packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts (hunk 5 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts]: Which others do you mean?

File: tests/angular_devkit/build_angular/test-differential-loading/tsconfig.app.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (alan-agius4) [inline tests/angular_devkit/build_angular/test-differential-loading/tsconfig.app.json]: I am not quite sure why we need to add all these tsconfig.json files.

The large spec, should cover an existing application that we have a large spec.

File: packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts (hunk 5 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts]: This one https://github.com/angular/angular-cli/pull/13993/files#diff-7379beedd1dbd6738b8da9124058d8b2

File: bin/gen-browser-infos.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+const caniuse = require('caniuse-api');
```
Comment (alan-agius4) [inline bin/gen-browser-infos.js]: What is the usage of this file?

Comment (alan-agius4) [inline bin/gen-browser-infos.js]: how is this hooked in the build system?

File: bin/gen-browser-infos.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+const fs = require('fs');
+
+const path =
+  '../packages/angular_devkit/build_angular/src/differential-loading/es6-module-support.json';
+const contents = JSON.stringify(caniuse.getSupport('es6-module'));
+
+fs.writeFileSync(path, contents, { encoding: 'UTF-8' });
+
```
Comment (alan-agius4) [inline bin/gen-browser-infos.js]: Nit: remove extra breaklines

File: packages/angular_devkit/build_angular/package.json [hunk 9]
```diff
@@ -13,6 +13,7 @@
     "@ngtools/webpack": "0.0.0",
     "ajv": "6.10.0",
     "autoprefixer": "9.4.9",
+    "browserslist": "^4.4.2",
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/package.json]: Direct dependencies should be pinned.

File: packages/angular_devkit/build_angular/package.json [hunk 10]
```diff
@@ -26,24 +27,25 @@
...
+    "source-map-support": "0.5.11",
+    "speed-measure-webpack-plugin": "1.3.1",
     "stats-webpack-plugin": "0.7.0",
     "style-loader": "0.23.1",
     "stylus": "0.54.5",
     "stylus-loader": "3.0.2",
-    "tree-kill": "1.2.1",
     "terser-webpack-plugin": "1.2.3",
+    "tree-kill": "1.2.1",
+    "typescript": "^3.3.4000",
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/package.json]: TypeScript should be a peerDependency and we need to also support older versions of TS
```
    "typescript": ">=2.7 < 3.4",
```

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 4 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: What be good to do some internal interfaces, for type checking.

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  } else if (version1Tupel[1] < version2Tupel[1]) {
+    return -1;
+  } else if (version1Tupel[1] > version2Tupel[1]) {
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+function toVersionTupel(version: string): [number, number] {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Kindly confused what this tuples means. What is version `[0]` and what is `[1]`?

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 4 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: +1

File: bin/gen-browser-infos.js (hunk 7 above)
Comment (mgechev) [reply bin/gen-browser-infos.js]: 👍

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 2 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Coincidental cohesion is not ideal. I like the current name. Maybe `_` instead of `-` so we can be consistent?

File: packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts (hunk 5 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts]: Ah, I see. It's about merging the `spec` with the `large_spec` file, right? 

Currently, `spec` does everything we can do in memory. If I need to access the hard disk, I'm using `large_spec`. Isn't this the intended way?

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+export function isEs5BrowserSupportNeeded(
+  tsConfigFileName: string | null, queries: string[] | undefined = undefined) {
+
+  const options: browserslist.Options = {};
+  if (tsConfigFileName) {
+    options.path = path.dirname(tsConfigFileName);
+  }
+
+  const browsersList: string[] = browserslist(queries, options);
+
```
Comment (mgechev) [inline packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: Extra space

File: tests/angular_devkit/build_angular/test-differential-loading/tsconfig.app.json (hunk 6 above)
Comment (mgechev) [reply tests/angular_devkit/build_angular/test-differential-loading/tsconfig.app.json]: Below we have an example of a tsconfig cycle, this is probably not covered by our existing app?

Maybe `readTsconfig` covers this?

File: packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts (hunk 5 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts]: Typically test large are used to test the full architect builder and not a unit.

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 2 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: That is true, though I am not sure about the file name per see, since the exported methods here, are really not related to 'differential loading enabled' per see.

Such as `getScriptTarget`, also, I won’t name a file with an `is` prefix in that case. Unless this file has a single responsibility. Though still naming a module with an `is` prefix doesn’t seems right to me. imho.

Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: For me, both is ok. The only function we need to export is `isDifferentialLoadingNeeded`. The others are only exported for testability.

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 3 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: +1

File: packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts (hunk 5 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/test/differential-loading/is-differential-loading-needed_spec_large.ts]: Ok, I see. In this case, merging makes sense.

File: tests/angular_devkit/build_angular/test-differential-loading/tsconfig.app.json (hunk 6 above)
Comment (manfredsteyer) [reply tests/angular_devkit/build_angular/test-differential-loading/tsconfig.app.json]: Yes, this was the idea. As I was not aware of `readTsconfig`, I wrote this by my self. And hence, I wanted to test the logic with a tsconfig pointing to ES5, a tsconfig pointing to ES2015, a tsconfig w/o a target pointing to another tsconfig with a extends cycle. 

Now, as I'm using `readTsconfig` this does not make sense anymore as we can assume that this function which is part of typescript does its job.

File: bin/gen-browser-infos.js (hunk 7 above)
Comment (manfredsteyer) [reply bin/gen-browser-infos.js]: We can do this. I was not sure if we should add this 1.x MB lib even though we just need this simple piece of data.

File: bin/gen-browser-infos.js (hunk 8 above)
Comment (manfredsteyer) [reply bin/gen-browser-infos.js]: +1

File: packages/angular_devkit/build_angular/package.json (hunk 9 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/package.json]: +1

File: packages/angular_devkit/build_angular/package.json (hunk 10 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/package.json]: +1

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 11 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: It's about the major and the minor part of the version. And about comparing two versions with a major and a minor. I can replace the tuple with a class. 

Or do we already have such a function somewhere? 

In this case, reusing it would be the way to go, I guess ...

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 12 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: What do you mean?

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 3 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: I'll rename the variable to `browserslistQuery`.

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 2 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: it's now called util

File: packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts (hunk 11 above)
Comment (clydin) [reply packages/angular_devkit/build_angular/src/differential-loading/is-differential-loading-needed.ts]: The `semver` package can be used and is already a dependency.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+function toVersionTupel(version: string): [number, number] {
+  const temp = version.split('.');
+  if (temp.length === 1) {
+    temp[1] = '0';
+  }
+
+  return [parseInt(temp[0]), parseInt(temp[1])];
+}
+
+function parseBrowserVersionTuple(raw: string): [string, string] {
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: Doesn't appear to be used?

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as browserslist from 'browserslist';
+import * as caniuse from 'caniuse-api';
+import * as path from 'path';
+import { ScriptTarget } from 'typescript';
+import { readTsconfig } from '../angular-cli-files/utilities/read-tsconfig';
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: This will most likely be removed in an upcoming refactoring/cleanup.  You can use `ts.getParsedCommandLineOfConfigFile` directly.

File: tests/angular_devkit/build_angular/test-differential-loading/index.html [hunk 15]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html><html lang="en"><head>
```
Comment (clydin) [inline tests/angular_devkit/build_angular/test-differential-loading/index.html]: These HTML files don't appear to be used.

**Comment by alan-agius4** [comment]
Can you also rebase?

File: packages/angular_devkit/build_angular/package.json [hunk 16]
```diff
@@ -26,55 +29,58 @@
     "loader-utils": "1.2.3",
     "mini-css-extract-plugin": "0.5.0",
     "minimatch": "3.0.4",
-    "parse5": "4.0.0",
     "opn": "5.4.0",
+    "parse5": "4.0.0",
```
[the remark this answers is not in the document]
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/package.json]: `parse5` is used in `index-html-webpack-plugin.ts`

[the remark this answers is not in the document]
Comment (clydin) [reply packages/angular_devkit/build_angular/package.json]: yes.  however, it is being moved around; along with several other entries (some with version number changes as well)

File: packages/angular_devkit/build_angular/package.json [hunk 17]
```diff
@@ -14,6 +14,8 @@
     "@ngtools/webpack": "0.0.0",
     "ajv": "6.10.0",
     "autoprefixer": "9.5.0",
+    "browserslist": "^4.5.2",
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/package.json]: pin direct dependencies please

Comment (alexeagle) [reply packages/angular_devkit/build_angular/package.json]: alternatively, consider vendoring them into our repo so we don't add more transitive dependencies

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  tsConfigFileName: string | null, browserslistQuery: string[] | undefined = undefined): boolean {
+
+  const neededFeature = 'es6-module';
+
+  const options: browserslist.Options = {};
+  if (tsConfigFileName) {
+    options.path = path.dirname(tsConfigFileName);
+  }
+
+  const browsersList: string[] = browserslist(browserslistQuery, options);
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: We don't need to use `browserslist` ourselves `caniuse.isSupported`, supports `browserslistQuery`.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+
+export function isDifferentialLoadingNeeded(
+  tsConfigFileName: string, browserslistQuery: string[] | undefined = undefined): boolean {
+  const es5SupportNeeded = isEs5BrowserSupportNeeded(tsConfigFileName, browserslistQuery);
+  const target = getTargetVersion(tsConfigFileName);
+
+  return es5SupportNeeded && target !== ts.ScriptTarget.ES5;
+}
+
+export function getTargetVersion(tsConfigFileName: string): ts.ScriptTarget {
```
Comment (alexeagle) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: is it possible to reference a `ts.Program` already at this point, so we don't have to worry about things like whether we found the right `tsconfig` file?

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+    // fall back to ES5
+    return ts.ScriptTarget.ES5;
+  }
+
+  return target;
+}
+
+export function isEs5BrowserSupportNeeded(
+  tsConfigFileName: string | null, browserslistQuery: string[] | undefined = undefined): boolean {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: Wouldn't these browser list be coming from `browserslist` file which is being read below?

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 19 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: We're using this to get the `target` output. We determine the maximum version (i.e. the `script[module]` one) from this property.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 20 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: Absolutely. I'm currently using the option to pass in a `browserslistQuery` instead of reading the `browserslist` file for testing.

File: packages/angular_devkit/build_angular/package.json (hunk 17 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/package.json]: I've pinned it now

File: packages/angular_devkit/build_angular/package.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/package.json]: It's now updated and it has a minimal amount of changes.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 19 above)
Comment (alexeagle) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: sure, but how does the user of this API know how to locate the correct tsconfig file if there are several in the project?

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 20 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: Ummm, probably we need to remove these before merging this PR.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 19 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: The user of this API will be the browser builder which finds the tsconfig.json to use in the `angular.json`. It can also get overruled by a command line parameter.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 20 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: I've changed it now. The unit tests read the browserlistQuerys now from the hard disk. Hence, there is no need for this `browserslistQuery` parameter anymore.

File: packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts [hunk 21]
```diff
@@ -0,0 +1,75 @@
+/**
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: I think there is a bit of misunderstanding here, specs are usefully next to the feature file. These specs here are usefully used to test the architect and named `large`.

These are more of an integration test rather than a unit. Unit tests should reside next to the feature.

Also, we should not interact with the File System directly, we should use the Abstraction layer as this would reduce a couple of things.

1) Makes mocking easier.
2) No need to create a physical file for each test. 
3) Have the mocks as part of the spec file hence makes the test more readable.

```
const workspaceRoot = path.join(
  devkitRoot,
  'tests/angular_devkit/build_angular/test-differential-loading/');
const host = new TestProjectHost(normalize(workspaceRoot));

describe('differential loading', () => {
  it('detects the need for ES5 if IE 9-11 has to be supported', () => {
    const path = normalize('browerlist');
    host.writeMultipleFiles({
        path: 'last 1 chrome version'
    });
    const es5SupportNeeded = isEs5BrowserSupportNeeded(workspaceRoot, path);
    expect(es5SupportNeeded).toBe(true);
  });
});
```

The `host` itself has a lot for method, that can help you with transforming a file, overriding it's content, delete it etc..., the important one is that it will never override the file on the hard drive.

Comment (alan-agius4) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: Umm actually `TestProjectHost` seems to be deprecated, @clydin what's the new preferred way to approach this?

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 19 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: If the user of this is browser builder, we already know the script target and if es5Support is needed or not. As there is where we modify the mainFields to include es2015 or not. So why do we need to get this again?

When you say that it can be overrules by a command line parameter, can you please elaborate? Wouldn't such flag be in the browser builder?

I think in general there is a bit of a lack of understanding from me on how everything will be glued together.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+    // fall back to ES5
+    return ts.ScriptTarget.ES5;
+  }
+
+  return target;
+}
+
+export function isEs5BrowserSupportNeeded(
+  browserslistPath: string, browserslistConfigFile: string | undefined = undefined): boolean {
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: How would we know the `browserslistConfigFile` from the CLI? this can be anything which browserlist supports. While at the moment in the CLI we generate an emotionless `browserlist`, the users are free to change this as long as browserlist can read it? 

Hence, why not simply provide the workspacePath and let `browserslist` resolve it''s configuration file? Similar to how it's currently done by autoprefixer in the style config.

`browserslist` configuration can be in various files such as;

1. package.json
2. .browserslistrc
3. browserslist

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 19 above)
Comment (mgechev) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: @alan-agius4 what you're saying sounds right and aligns with the design. On `browserlist` depends if we should generate ES5 bundles or not.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
+/**
```
Comment (alan-agius4) [inline packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: To determine whether deferential loading is needed it should be simple;

```ts
// In browser builder we already know if the it supports `supportES2015` as we use this to populate the correct `mainFields`
export function isDifferentialLoadingNeeded(projectRoot: string, supportES2015: boolean): boolean {
  const browsersList: string[] = browserslist(undefined, {path: projectRoot});
  return supportES2015 && !caniuse.isSupported('es6-module', browsersList.join(', '));
}
```

or
```ts
// In browser builder we already know the target
export function isDifferentialLoadingNeeded(projectRoot: string, target: ts.ScriptTarget): boolean {
  const supportES2015 = target !== projectTs.ScriptTarget.ES3 && target !== projectTs.ScriptTarget.ES5;
  const browsersList: string[] = browserslist(undefined, {path: projectRoot});
  return supportES2015 && !caniuse.isSupported('es6-module', browsersList.join(', '));
}
```

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 22 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: This should be the default, I agree. The current design doc also specifies that we should be able to define one using a parameter (command line parameter when calling ng build or a parameter within angular.json). I think, this would come in-handy b/c so one could build versions for different targets.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 23 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: I see your point. Thank you!

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 19 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: Ok, I see what you mean. THX!

File: packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts (hunk 21 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: Still pending.

File: packages/angular_devkit/build_angular/src/differential-loading/utils.ts (hunk 22 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: At the moment, having different multiple browserlist configs is not supported by autoprefixer. it only supports either resolving the browserlist file as per it's internal resolution strategy or an array of browsers to support.

My point is that if we are going to provide a way for users to change the browserlist with a configuration parameter.  We need to change the logic in a couple of builders as this would also determine which prefixes would need to be used during server builder.

IMHO, if such approach will need to happen it shouldn't be the responsibility of the differential loader utils and it should be somewhere more centralised:

In general I'd like to avoid having to parse the same file multiple times.
-> parse browser list config as required
--> pass this down the differential loader
--> pass this down to the webpack styles configuration.

That said, I am not seeing what would be the users case of building the application several times using different browser versions. (I will check the design doc)

Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: Ok, so we leave this away for now, right?

Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/src/differential-loading/utils.ts]: I removed this param now

File: packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts (hunk 21 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: I've tried this. While this totally makes sense, it seems that in this very case, using the virtual filesystem does not work. The reason is that browserslist has its own machanism for finding and loading the `browswerslist` config. This one isn't using the virtual filesystem.

Comment (alan-agius4) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: This should work, as the TestProjectHost writes files to disks.

See below;
Screenshot 2019-04-02 at 17 45 13



If you need a hand let me know.

Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: It seems to be a Windows-related issue: `host.root()` returns sth like `/c/my-dir` instead of `c:/my-dir`. The forward slashes are no issue here.

Comment (alan-agius4) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: Yeah, that is expected if you pass a devkit path directly to the method yourself that is using FS.

Devkit paths are always posix normalised, which means on windows `c:\\` will be `/c/` which means that any interaction with FS will fail on windows. Devkit paths/methods cannot be used with direct interaction with fs. 

The projectRoot, is normalised is passed from `wco`. See https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/build_angular/src/browser/index.ts#L127

The safest and cleaner approach would be changing the `projectRoot` to Path and using `getSystemPath`

```
export function isDifferentialLoadingNeeded(
  projectRoot: Path,
  target: ts.ScriptTarget): boolean {

   const supportES2015 = target !== ts.ScriptTarget.ES3 && target !== ts.ScriptTarget.ES5;
  const browsersList: string[] = browserslist(
    undefined, {
      path: getSystemPath(projectRoot),
    });

   return supportES2015 && !caniuse.isSupported('es6-module', browsersList.join(', '));
}
```

**Comment by alan-agius4** [comment]
@manfredsteyer is this absolute in favor of the other PR or not?

File: packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts (hunk 21 above)
Comment (manfredsteyer) [reply packages/angular_devkit/build_angular/test/differential-loading/differential_loading_spec.ts]: Thanks!

**Comment by manfredsteyer** [comment]
> in favor of the



> @manfredsteyer is this absolute in favor of the other PR or not?

Yes, it is @alan-agius4
